File: flowstore/flow.go
package flowstore

import (
	"fmt"
	"time"

	"github.com/c360/semstreams/pkg/errs"
	"github.com/c360/semstreams/types"
)

// Flow represents a visual flow definition with metadata and canvas layout
type Flow struct {
	// Identity
	ID          string `json:"id"`
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`

	// Version for optimistic concurrency control
	Version int64 `json:"version"`

	// Canvas layout
	Nodes       []FlowNode       `json:"nodes"`
	Connections []FlowConnection `json:"connections"`

	// Runtime state
	RuntimeState RuntimeState `json:"runtime_state"`
	DeployedAt   *time.Time   `json:"deployed_at,omitempty"`
	StartedAt    *time.Time   `json:"started_at,omitempty"`
	StoppedAt    *time.Time   `json:"stopped_at,omitempty"`

	// Audit
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
	CreatedBy    string    `json:"created_by,omitempty"`
	LastModified time.Time `json:"last_modified"`
}

// FlowNode represents a component instance on the canvas
type FlowNode struct {
	ID            string              `json:"id"`             // Unique instance ID
	ComponentID   string              `json:"component_id"`   // Factory name (e.g., "udp", "graph-processor")
	ComponentType types.ComponentType `json:"component_type"` // Category (input/processor/output/storage/gateway)
	Name          string              `json:"name"`           // Instance name
	Position      Position            `json:"position"`       // Canvas coordinates
	Config        map[string]any      `json:"config"`         // Component configuration
}

// FlowConnection represents a connection between two component ports
type FlowConnection struct {
	ID           string `json:"id"`
	SourceNodeID string `json:"source_node_id"`
	SourcePort   string `json:"source_port"`
	TargetNodeID string `json:"target_node_id"`
	TargetPort   string `json:"target_port"`
}

// Position represents canvas coordinates for a node
type Position struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

// RuntimeState represents the deployment and execution state of a flow
type RuntimeState string

// RuntimeState constants define the lifecycle states of a flow:
//   - StateNotDeployed: Flow exists but has never been deployed
//   - StateDeployedStopped: Flow deployed to config but not running
//   - StateRunning: Flow is actively processing messages
//   - StateError: Flow encountered an error during deployment/execution
const (
	StateNotDeployed     RuntimeState = "not_deployed"
	StateDeployedStopped RuntimeState = "deployed_stopped"
	StateRunning         RuntimeState = "running"
	StateError           RuntimeState = "error"
)

// Validate checks if the flow is valid for deployment
func (f *Flow) Validate() error {
	// Validate flow-level fields
	if f.ID == "" {
		return errs.WrapInvalid(fmt.Errorf("flow ID cannot be empty"), "flowstore", "Validate", "validation failed")
	}
	if f.Name == "" {
		return errs.WrapInvalid(fmt.Errorf("flow name cannot be empty"), "flowstore", "Validate", "validation failed")
	}

	// Validate runtime state
	validStates := map[RuntimeState]bool{
		StateNotDeployed:     true,
		StateDeployedStopped: true,
		StateRunning:         true,
		StateError:           true,
	}
	if !validStates[f.RuntimeState] {
		return errs.WrapInvalid(
			fmt.Errorf("invalid runtime state: %s", string(f.RuntimeState)),
			"flowstore", "Validate", "runtime state validation failed")
	}

	// Validate nodes
	nodeIDs := make(map[string]bool)
	for i, node := range f.Nodes {
		if node.ID == "" {
			return errs.WrapInvalid(
				fmt.Errorf("node at index %d has empty ID", i),
				"flowstore", "Validate", "node ID validation failed")
		}
		if node.ComponentID == "" {
			return errs.WrapInvalid(
				fmt.Errorf("node '%s' has empty component_id", node.ID),
				"flowstore", "Validate", "node component_id validation failed")
		}
		if node.ComponentType == "" {
			return errs.WrapInvalid(
				fmt.Errorf("node '%s' has empty component_type", node.ID),
				"flowstore", "Validate", "node component_type validation failed")
		}
		if node.Name == "" {
			return errs.WrapInvalid(
				fmt.Errorf("node '%s' has empty name", node.ID),
				"flowstore", "Validate", "node name validation failed")
		}

		// Check for duplicate node IDs
		if nodeIDs[node.ID] {
			return errs.WrapInvalid(
				fmt.Errorf("duplicate node ID: %s", node.ID),
				"flowstore", "Validate", "duplicate node ID detected")
		}
		nodeIDs[node.ID] = true
	}

	// Validate connections
	for i, conn := range f.Connections {
		if conn.ID == "" {
			return errs.WrapInvalid(
				fmt.Errorf("connection at index %d has empty ID", i),
				"flowstore", "Validate", "connection ID validation failed")
		}
		if conn.SourcePort == "" {
			return errs.WrapInvalid(
				fmt.Errorf("connection '%s' has empty source port", conn.ID),
				"flowstore", "Validate", "connection source port validation failed")
		}
		if conn.TargetPort == "" {
			return errs.WrapInvalid(
				fmt.Errorf("connection '%s' has empty target port", conn.ID),
				"flowstore", "Validate", "connection target port validation failed")
		}

		// Validate node references
		if !nodeIDs[conn.SourceNodeID] {
			return errs.WrapInvalid(
				fmt.Errorf("connection '%s' references non-existent source node: %s", conn.ID, conn.SourceNodeID),
				"flowstore", "Validate", "connection source node validation failed")
		}
		if !nodeIDs[conn.TargetNodeID] {
			return errs.WrapInvalid(
				fmt.Errorf("connection '%s' references non-existent target node: %s", conn.ID, conn.TargetNodeID),
				"flowstore", "Validate", "connection target node validation failed")
		}
	}

	return nil
}

File: ui/src/lib/types/flow.ts
// Type definitions for Flow entities
// Matches backend schema from pkg/flowstore/flow.go

import type { RuntimeState, SaveStatus, UserPreferences } from "./ui-state";
import type { ValidationState as PortValidationState } from "./port";

// Re-export for convenience
export type { RuntimeState, SaveStatus, UserPreferences };
export { DEFAULT_PREFERENCES } from "./ui-state";

export interface Flow {
  // Identity
  id: string;
  name: string;
  description?: string;

  // Version for optimistic concurrency
  version: number;

  // Canvas layout (backend calls them 'nodes' not 'components')
  nodes: FlowNode[];
  connections: FlowConnection[];

  // Runtime state
  runtime_state: RuntimeState;
  deployed_at?: string; // ISO 8601, optional
  started_at?: string; // ISO 8601, optional
  stopped_at?: string; // ISO 8601, optional

  // Audit
  created_at: string; // ISO 8601
  updated_at: string; // ISO 8601
  created_by?: string;
  last_modified: string; // ISO 8601
}

// Helper type for UI state (not from backend)
export interface FlowMetadata {
  saveStatus: SaveStatus;
}

// Backend schema: FlowNode
export interface FlowNode {
  id: string;
  component: string; // References ComponentType.id (e.g., "json_filter")
  type: string; // Component type for colors: "input" | "output" | "processor" | "gateway" | "storage"
  name: string;
  position: Position;
  config: Record<string, unknown>; // Component-specific config
}

// UI-only: Extended node with health status
export interface ComponentInstance extends FlowNode {
  health: ComponentHealth;
}

export interface Position {
  x: number;
  y: number;
}

export interface ComponentHealth {
  status: HealthStatus;
  errorMessage?: string;
  lastUpdated: string; // ISO 8601
}

export type HealthStatus = "healthy" | "degraded" | "unhealthy" | "not_running";

/**
 * Connection source type
 * Distinguishes auto-discovered connections (from FlowGraph) vs manually created
 */
export type ConnectionSource = "auto" | "manual";

// Backend schema: FlowConnection (EXTENDED for visual component wiring)
export interface FlowConnection {
  id: string;
  source_node_id: string;
  source_port: string;
  target_node_id: string;
  target_port: string;

  // NEW: Connection source tracking (auto-discovered vs manual)
  // Optional for backward compatibility with existing code
  source?: ConnectionSource;

  // NEW: UI validation state (not persisted to backend)
  validationState?: PortValidationState;
  validationMessage?: string;
}

// UI-only: Extended connection with validation and metrics
export interface Connection extends FlowConnection {
  pattern: InteractionPattern;
  validationError?: string;
  metrics?: ConnectionMetrics;
}

export type InteractionPattern = "stream" | "request" | "watch" | "network";

export interface ConnectionMetrics {
  messageCount: number;
  throughputRate: number; // messages/second
  lastActivity: string; // ISO 8601
}

// ============================================================================
// Phase 3: Connection Visual States (Spec 014)
// ============================================================================

/**
 * Visual styling for connection lines (XYFlow edges)
 *
 * Based on pattern type (stream/request/watch) and validation state.
 * Includes line patterns, colors, error indicators, and tooltips.
 *
 * @see specs/014-flow-ux-port/data-model.md
 */
export interface ConnectionVisualState {
  /** Connection ID (XYFlow edge ID) */
  connectionId: string;

  /** Line style based on pattern type */
  linePattern: "solid" | "dashed" | "dotted";

  /** Line color based on validation state */
  color: string;

  /** Validation state */
  validationState: "valid" | "error" | "warning";

  /** SVG stroke-dasharray value */
  strokeDasharray: string;

  /** Z-index for layering (selected connections on top) */
  zIndex: number;

  /** Optional error icon to display on line */
  errorIcon?: string;

  /** Tooltip content for connection */
  tooltipContent?: string;
}

/**
 * Real-time visual feedback during connection creation
 *
 * Shows port compatibility in real-time as user drags from source port.
 * Compatible ports show green highlight, incompatible show red indicator.
 *
 * @see specs/014-flow-ux-port/data-model.md
 */
export interface CompatibilityFeedback {
  /** Currently dragged source port ID */
  sourcePortId: string;

  /** Target port being hovered over (if any) */
  targetPortId?: string;

  /** Compatibility status */
  compatibility: "compatible" | "incompatible" | "unknown";

  /** Visual indicator to display */
  indicator: "green-highlight" | "red-indicator" | "none";

  /** Reason for incompatibility (if any) */
  incompatibilityReason?: string;

  /** CSS classes for visual feedback */
  feedbackClasses: string[];
}

/**
 * XYFlow edge with visual enhancements
 *
 * Extends XYFlow edge data with connection visual state.
 * Used for pattern-specific styling and error/warning indicators.
 *
 * @see specs/014-flow-ux-port/data-model.md
 */
export interface XYFlowEdgeWithVisuals {
  id: string;
  source: string;
  target: string;
  sourceHandle?: string;
  targetHandle?: string;
  type?: string;
  data?: {
    /** Connection visual state */
    visualState?: ConnectionVisualState;

    /** Pattern type (from backend validation) */
    patternType?: "stream" | "request" | "watch";

    /** Source (auto-discovered vs manual) */
    connectionSource?: "auto" | "manual";
  };
}
